notification.c

```diff
@@ -352,15 +357,46 @@ void notification_execute_binding(struct mako_notification *notif,
 		struct mako_action *action;
 		wl_list_for_each(action, &notif->actions, link) {
 			if (strcmp(action->key, DEFAULT_ACTION_KEY) == 0) {
-				char *activation_token = create_xdg_activation_token(
-					ctx->surface, ctx->seat, ctx->serial);
+				char *activation_token = NULL;
+				if (ctx != NULL) {
+					activation_token = create_xdg_activation_token(ctx->surface,
+						ctx->seat, ctx->serial);
+				}
 				notify_action_invoked(action, activation_token);
 				free(activation_token);
 				break;
 			}
 		}
 		close_notification(notif, MAKO_NOTIFICATION_CLOSE_DISMISSED);
 		break;
+	case MAKO_BINDING_EXEC:
+		assert(binding->command != NULL);
+		pid_t pid = fork();
+		if (pid < 0) {
+			perror("fork failed");
+			break;
+		} else if (pid == 0) {
+			// Double-fork to avoid SIGCHLD issues
+			pid = fork();
+			if (pid < 0) {
+				perror("fork failed");
+				_exit(1);
+			} else if (pid == 0) {
+				char id_str[32];
+				snprintf(id_str, sizeof(id_str), "%" PRIu32, notif->id);
+				setenv("id", id_str, 1);
+
+				char *const argv[] = { "sh", "-c", binding->command, NULL };
+				execvp("sh", argv);
+				perror("exec failed");
```

Comment:
Should probably reset errno first just to be safe since there's no other error check like with the forks.

Reply:
From the exec man page:

> The exec() functions return only if an error has occurred.  The return value is -1, and errno is set to indicate the error.

Reply:
Ah, duh.